x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportRepositoryStatsAction.java

```diff
@@ -0,0 +1,89 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.repositories.RepositoriesService;
+import org.elasticsearch.repositories.Repository;
+import org.elasticsearch.repositories.RepositoryStats;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+public class TransportRepositoryStatsAction extends TransportNodesAction<
+    RepositoryStatsRequest,
+    RepositoryStatsResponse,
+    RepositoryStatsNodeRequest,
+    RepositoryStatsNodeResponse> {
+
+    private final RepositoriesService repositoriesService;
+    private final XPackLicenseState licenseState;
+
+    @Inject
+    public TransportRepositoryStatsAction(
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        RepositoriesService repositoriesService,
+        XPackLicenseState licenseState
+    ) {
+        super(
+            RepositoryStatsAction.NAME,
+            threadPool,
+            clusterService,
+            transportService,
+            actionFilters,
+            RepositoryStatsRequest::new,
+            RepositoryStatsNodeRequest::new,
+            ThreadPool.Names.SAME,
+            RepositoryStatsNodeResponse.class
+        );
+        this.repositoriesService = repositoriesService;
+        this.licenseState = Objects.requireNonNull(licenseState);
+    }
+
+    @Override
+    protected RepositoryStatsResponse newResponse(
+        RepositoryStatsRequest request,
+        List<RepositoryStatsNodeResponse> nodes,
+        List<FailedNodeException> failures
+    ) {
+        return new RepositoryStatsResponse(clusterService.getClusterName(), nodes, failures);
+    }
+
+    @Override
+    protected RepositoryStatsNodeRequest newNodeRequest(RepositoryStatsRequest request) {
+        return new RepositoryStatsNodeRequest(request.getRepository());
+    }
+
+    @Override
+    protected RepositoryStatsNodeResponse newNodeResponse(StreamInput in) throws IOException {
+        return new RepositoryStatsNodeResponse(in);
+    }
+
+    @Override
+    protected RepositoryStatsNodeResponse nodeOperation(RepositoryStatsNodeRequest request, Task task) {
+        SearchableSnapshots.ensureValidLicense(licenseState);
+        if (clusterService.localNode().isMasterNode() == false && clusterService.localNode().isDataNode() == false) {
+            return new RepositoryStatsNodeResponse(clusterService.localNode(), RepositoryStats.EMPTY_STATS);
+        }
+        final Repository repository = repositoriesService.repository(request.getRepository());
```

Comment:
Should we catch the repository missing exception here - in case the repository is not yet known by this node - and return empty stats too?

Reply:
I would like to avoid that as it would then silently provide partial stats